Fix dotnet add package doesn't work with relative path source or when version is not specified

erdembayar opened the pull request:
## Bug

Fixes: [Issue#5127](https://github.com/NuGet/Home/issues/5127)
Regression: No  
* Last working version:   
* How are we preventing it in future:   

## Fix

Details:  Currently if user add package with dotnet tool from cli with custom source feed(local, remote) as argument then it doesn't work unless user supply exact version. Also this behavior changes between V3 and V2 format sources.

Below is whole steps:
Create a folder named 'Domain'
Create a folder named 'App'

Inside of Domain, run these commands:

```
dotnet new classlib
dotnet restore
dotnet pack --include-symbols --output ..\Mysource
Inside of App, run these commands:

dotnet new classlib
dotnet restore
dotnet add package Domain -s ..\Mysource 

```

Don't forget to delete Domain package from local cache after successful operation, otherwise always success regardless of input next time.

**V3 source:**
Absolute path:
`dotnet add package Domain3 -s C:\NuGetProj\IssueRepro\5127\Mypackages3 -v 1.0.0` works

`dotnet add package Domain3 -s C:\NuGetProj\IssueRepro\5127\Mypackages3`  error: `There are no versions available for the package 'Domain3'.` (Fixed)

Relative path:
`dotnet add package Domain3 -s ..\Mypackages3 -v 1.0.0` works
`dotnet add package Domain3 -s ..\Mypackages3` error: `There are no versions available for the package 'Domain3'.`
(Fixed)

**V2 source:**
Absolute path:
`dotnet add package Domain2 -s C:\NuGetProj\IssueRepro\5127\Mysource2 -v 1.0.0` works.

`dotnet add package Domain2 -s C:\NuGetProj\IssueRepro\5127\Mysource2` error: `There are no versions available for the package 'Domain2'.`    (Fixed)

Relative path:
`dotnet add package Domain2 -s ..\Mysource2 -v 1.0.0` error: `NU1101: Unable to find package Domain2. No packages exist with this id in source(s): ..\Mysource2, nuget.org`  
error: `Package 'Domain2' is incompatible with 'all' frameworks in project 'C:\NuGetProj\IssueRepro\5127\App2\App2.csproj'.`
(Fixed)

`dotnet add package Domain2 -s ..\Mysource2`  error: `There are no versions available for the package 'Domain2'.`
(Fixed)

**Please note**:  To make this code consistent with other `dotnet command` with user supplied custom source feeds in this fix also replaces dgSpec sources with user supplied custom source feeds.

For example below replaces source feeds in dgSpec file;
`dotnet restore /p:restoreSources="C:\NuGetProj\IssueRepro\5127\Mypackages3;C:\NuGetProj\IssueRepro\5127\Mysource2"`  

## Testing/Validation
Tests Added: Yes  
Reason for not adding tests:  
Validation:  Manual from VS and patched sdk5.

erdembayar committed:
Fix 'dotnet add package' with user supplied source feed works if no version is specified.

erdembayar committed:
Prevent from user enter duplicate source from cli.

zkat reviewed (commented):
This eyeballs fine to me. I would like to see tests specific to this scenario before approving, though.

erdembayar commented:
> This eyeballs fine to me. I would like to see tests specific to this scenario before approving, though.

@zkat 
Sure, I'll add unit test by tomorrow.

nkolev92 commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/AddPackageReferenceCommandRunner.cs:
```diff
@@ -105,13 +105,13 @@ public async Task<int> ExecuteCommand(PackageReferenceArgs packageReferenceArgs,
             if (packageReferenceArgs.NoVersion)
             {
-                var latestVersion = await GetLatestVersionAsync(originalPackageSpec, packageReferenceArgs.PackageId, packageReferenceArgs.Logger, packageReferenceArgs.Prerelease);
+                var latestVersion = await GetLatestVersionAsync(originalPackageSpec, packageReferenceArgs.Sources, packageReferenceArgs.PackageId, packageReferenceArgs.Logger, packageReferenceArgs.Prerelease);
```
Did you test this? :) 

Pretty sure this is not enough, as the installation is 2 step, once to get the version, once to actually download the package, and the additional source is only used for the version calculation.

erdembayar replied:
> Did you test this? :)
> 
> Pretty sure this is not enough, as the installation is 2 step, once to get the version, once to actually download the package, and the additional source is only used for the version calculation.

Yes, manually I did test with both local and remote private feeds.
1. With my change it get's latest version from additional user source feeds.
2. After that this additional user source feeds are passed to runner job for actual restore.

Can you refer where is `dotnet pack ....` unit tests?

nkolev92 replied:
Search for PackageCommandTests...not sure why you need `Pack` in this case though.

erdembayar replied:
Added unit tests.

nkolev92 reviewed (commented):
You won't know if you are the correct thing if you don't test it.

erdembayar committed:
Add unit tests for 'dotnet pack Package' command with user supplied source feed.

erdembayar commented:
> This eyeballs fine to me. I would like to see tests specific to this scenario before approving, though.

@zkat 
Please check I added 2 unit test for successful paths. Can't think of actual fail case specific to this case.

erdembayar committed:
Fix test breaking on Linux.

erdembayar committed:
2nd attempt to fix unit test failing on Linux.

nkolev92 commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/XPlatTestUtils.cs:
```diff
@@ -429,5 +429,11 @@ public static string GetReferenceType(PackageType packageType)
             return referenceType;
         }
+
+        public static void DeleteDirectory(string path)
```
Look at LocalResourceUtils instead.

erdembayar replied:
Ok. Changed to that util

nkolev92 commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/XPlatAddPkgTests.cs:
```diff
@@ -745,6 +746,117 @@ public async Task AddPkg_ConditionalAddWithUserInputFramework_Success(string pac
         }
 
+        [Fact]
+        public async Task AddPkg_WithAdditionalSourceFeed_NoVersionSpecified_Success()
```
Tests are still failing. 

Please ping me when they are passing.

erdembayar replied:
All tests are passing now. 😄

erdembayar replied:
Unfortunately I pushed another small change right after it, so have to wait that one to clear.

erdembayar committed:
Fix failing Linux unit test3.

erdembayar committed:
Remove unneeded utility method.

nkolev92 commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/AddPackageCommandUtility.cs:
```diff
@@ -113,6 +114,32 @@ public static List<PackageSource> EvaluateSources(IList<PackageSource> requested
 
+                HashSet<string> currentPackagesSources = packageSources.Select(p => p.Source).ToHashSet(StringComparer.OrdinalIgnoreCase);
+
+                if (additionalSources?.Any() ?? false)
```
The following is easier to read: 

```cs
                if (additionalSources?.Any() == true)
```

It's also what's used heavily in `AddPackageReferenceCommandRunner`.

erdembayar replied:
Ok. Addressed.

nkolev92 commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/AddPackageCommandUtility.cs:
```diff
@@ -88,9 +88,10 @@ public static async Task<NuGetVersion> GetLatestVersionFromSourceAsync(PackageSo
         /// Returns the PackageSource with its credentials if available
         /// </summary>
         /// <param name="requestedSources">Sources to match</param>
+        /// <param name="additionalSources">Additional user supplied source feeds as argument</param>
```
I don't think they are supposed to be additional, but rather they are supposed to replace the existing sources.

erdembayar replied:
I believe it's treated as addition into existing sources user already setup on project. 
Here below screenshot case it already has nuget.org as source and also `Mysource` which is I passed as argument on [AddPackageReferenceCommandRunner.cs#L347](https://github.com/NuGet/NuGet.Client/blob/f9ad775ce70967f23338566b806b7bde1f623ec3/src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/AddPackageReferenceCommandRunner.cs#L347). Maybe user can use this one as alternative sourcefeed for just once during restore.

![image](https://user-images.githubusercontent.com/8766776/101226533-9d082000-3649-11eb-9d53-9cb86c014cad.png)

nkolev92 replied:
I don't think this is a technical question. 

`--source` for restore *completely* overwrites the sources. 

`dotnet add package` should work the same. 

There are docs that describe this behavior as well: https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-add-package .

erdembayar replied:
@nkolev92 
Can you suggest name for it? 
How about `Overriding source` or `Replacing source`, `New sources`?

nkolev92 replied:
The sources in the restore request are not used at all. 
The sources in the dg spec are used https://github.com/NuGet/NuGet.Client/blob/6562ef31f891c26981f5f5cc82f3fe5c21451eb9/src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/RestoreArgs.cs#L129-L144. 

Don't think you'd need to change it there. It'd be difficult to do without consequences, so I'd just change in the package spec that you are passing for restore.

nkolev92 replied:
Let's sync on Thursday if you are still stuck.

erdembayar replied:
@nkolev92 
Yes, can we talk 5-10 min today. Early is better.

erdembayar replied:
@nkolev92 
I tested example on [https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-add-package](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-add-package):
`dotnet add package Microsoft.AspNetCore.StaticFiles -s https://dotnet.myget.org/F/dotnet-core/api/v3/index.json
`

It doesn't change the source property on dgspec file. It's only used as temporary thing. It's not persisted, I can't say `--source`  completely overwrites the sources.
It's still 
```

        "sources": {
          "https://api.nuget.org/v3/index.json": {}
        },
```
Only sources listed in `dotnet nuget list source` or nuget.config go in there.
So when we use --source it's temporary source lets restore package from that source and persist copy of package in cache, but source details not saved. So better to call it `temp source`.

erdembayar replied:
It looks one problem with current behavior of `--source` we can't do repeatable restore/build if local cache is cleared even though we successfully added package from custom relative source unless that one is added into source list.

nkolev92 replied:
> It looks one problem with current behavior of --source we can't do repeatable restore

That's exactly why source is not supported in the PM UI & why you can't select a subset of sources when doing restore, it always uses them all. 

I think it's *ok* if you overwrite the source in this case. 
If the customer gets themselves in a non-repeatable position, it is what it is. 
They can already do that with /p:RestoreSources when they pass something different from what's specified in their nuget.config. 

> It doesn't change the source property on dgspec file. It's only used as temporary thing. It's not persisted, I can't say --source completely overwrites the sources.

I'm confused what this means. It doesn't work at all, it's disregarding it, which is what you are trying to fix.

erdembayar replied:
> I think it's _ok_ if you overwrite the source in this case.

@nkolev92 

I think overwriting is not good idea, maybe append is better. 

First of all let's run this:
`dotnet restore /p:restoreSources="C:\NuGetProj\IssueRepro\5127\Mypackages3;C:\NuGetProj\IssueRepro\5127\Mysource2"`

```
   Csproj file content: 

  <ItemGroup>
    <PackageReference Include="Domain3" Version="1.0.0" />
    <PackageReference Include="Domain2" Version="1.0.0" />
  </ItemGroup>
```

#### RestoreSuccessful
and source changes to below:

```
        "sources": {
          "C:\\NuGetProj\\IssueRepro\\5127\\Mypackages3": {},
          "C:\\NuGetProj\\IssueRepro\\5127\\Mysource2": {}
        },

```

### Another question is do we write relative path or absolute path in case  user passed relative path for source?

```
        "sources": {
          "..\Mysource2": {}
        },

```
or 

```
        "sources": {
          "C:\\NuGetProj\\IssueRepro\\5127\\Mysource2": {}
        },

```

### Overwrite

By default usually we have this source:
        "sources": {
          "https://api.nuget.org/v3/index.json": {}
        },

`dotnet add package Domain2 -s ..\Mysources2`

After overwrite we going to have 
```
        "sources": {
          "..\Mysources2": {},
        },
```

`dotnet add package Domain3 -s ..\Mypackages3`

After yet another overwrite we going to have 
```
        "sources": {
          "..\\Mypackages3": {},
        },
```

As you can see it's not great since every action overwrites the previous source.

### Append.

By default usually we have this source:
        "sources": {
          "https://api.nuget.org/v3/index.json": {}
        },

`dotnet add package Domain2 -s ..\Mysources2`

After append we going to have 
```
        "sources": {
          "..\Mysources2": {},
          "https://api.nuget.org/v3/index.json": {}
        },
```

`dotnet add package Domain3 -s ..\Mypackages3`

After append we going to have 
```
        "sources": {
          "..\Mysources2": {},
          "..\\Mypackages3": {},
          "https://api.nuget.org/v3/index.json": {}
        },
```

This looks better, because it looks similar to [RestoreSuccessful](#RestoreSuccessful).
Only issue is we have `"https://api.nuget.org/v3/index.json": {}` which is not in [RestoreSuccessful](#RestoreSuccessful).

Please give your thought?

nkolev92 commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/AddPackageCommandUtility.cs:
```diff
@@ -88,9 +88,10 @@ public static async Task<NuGetVersion> GetLatestVersionFromSourceAsync(PackageSo
         /// <param name="configFilePaths">Config to use for credentials</param>
         /// <returns>Return a list of package sources</returns>
-        public static List<PackageSource> EvaluateSources(IList<PackageSource> requestedSources, IList<string> configFilePaths)
+        public static List<PackageSource> EvaluateSources(IList<PackageSource> requestedSources, string[] additionalSources, IList<string> configFilePaths)
```
you added this new method, but you're not really testing the new capability.

erdembayar replied:
My 2 new unit tests with `no version specified` already cover this one. They come here to resolve the latest version for the package.

nkolev92 commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/XPlatAddPkgTests.cs:
```diff
@@ -745,6 +746,123 @@ public async Task AddPkg_ConditionalAddWithUserInputFramework_Success(string pac
+
+                if (!File.Exists(originalPackageXNupkg))
+                {
+                    // On linux it's created in different place.
```
why is it created in a different place?

erdembayar replied:
I don't know for now because I don't have any machine with Linux.
But I did create follow up [issue#10361](https://github.com/NuGet/Home/issues/10361)
Now it's not concern anymore for this PR.

nkolev92 commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/XPlatAddPkgTests.cs:
```diff
@@ -745,6 +746,123 @@ public async Task AddPkg_ConditionalAddWithUserInputFramework_Success(string pac
+                }
+
+                // We're testing case we have only Nupkg file created at top.
+                File.Move(originalPackageXNupkg, mySourcePackageXNupkg);
```
you created a v3 feed, which is consisted of more than just 1 file. Moving 1 file creates an inconsistent state in the original location.

nkolev92 replied:
look up v3 feeds in the sanity guide in the one note.

erdembayar replied:
What I'm trying to do is just get 1 Nupkg file, actually I don't need v3 feed.
Because `dotnet pack --include-symbols --output ..\Mysource` command just create 1 Nugkg file nothing else.
I saw [`Util.CreateTestPackage`](https://github.com/NuGet/NuGet.Client/blob/223189bbc21e8259e6771d363e69d37d9b10e693/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/Util.cs#L223) does actually that one but it's in different assembly and copying this method or referencing that assembly bring a lot of things.
So it's just work around to get just a nupkg file.

nkolev92 replied:
If you need V2, then use https://github.com/NuGet/NuGet.Client/blob/223189bbc21e8259e6771d363e69d37d9b10e693/test/TestUtilities/Test.Utility/SimpleTestSetup/SimpleTestPackageUtility.cs#L428-L433.

There are a lot of helper APIs we've built over time.

erdembayar replied:
Ok. Done. I'm still discovering them.

nkolev92 commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/XPlatAddPkgTests.cs:
```diff
@@ -745,6 +746,123 @@ public async Task AddPkg_ConditionalAddWithUserInputFramework_Success(string pac
+                // We're testing case we have only Nupkg file created at top.
+                File.Move(originalPackageXNupkg, mySourcePackageXNupkg);
+
+                // Remove packageXPath directory to prove we're not using nuget package structured folders.
```
I'm not sure what's happening here. You create a package in a location, but then you deleted it and moved it to another one? 
Why the delete?

erdembayar replied:
See [above comment](https://github.com/NuGet/NuGet.Client/pull/3783#discussion_r536443187).
Point is just prove I'm not using v3 feed make this command working, instead using just nupkg files.
Because `dotnet pack --include-symbols --output ..\Mysource` command create just a nupkg file, nothing else.
So it's just work around to get just a nupkg file.

erdembayar replied:
This is no more concern by moving to V2 feed.

nkolev92 commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/XPlatAddPkgTests.cs:
```diff
@@ -745,6 +746,123 @@ public async Task AddPkg_ConditionalAddWithUserInputFramework_Success(string pac
+        }
+
+        [Fact]
+        public async Task AddPkg_WithAdditionalSourceFeed_VersionSpecified_Success()
```
Same questions apply here as the above, it's a bit difficult to follow.

erdembayar replied:
I'll wait on your review on other unit test.

erdembayar replied:
Fixed it and added 2 more unit test for unhappy paths.

nkolev92 reviewed (commented):
Couple of more suggestions/questions.

erdembayar committed:
Address PR review by Nikolche.

erdembayar commented:
> Couple of more suggestions/questions.

@nkolev92 
I addressed your concern please check.

erdembayar committed:
Implemented V2 unget source feed instead of V3 source feed for unit tests.

erdembayar commented:
> Couple of more suggestions/questions.

@nkolev92 
It seems I misdiagnosed the problem. I'll change the PR.

nkolev92 commented:
I imagine the source problem still stands right? 
Which issue will this PR target?

erdembayar committed:
Correct conflicting FindLocalPackagesResourceV2Provider, FindLocalPackagesResourceV3Provider vs LocalV2FindPackageByIdResourceProvider, LocalV3FindPackageByIdResourceProvider.
Discovery of version and actual resolution of version had different logics.

erdembayar commented:
> I imagine the source problem still stands right?
> Which issue will this PR target?

Out of 8 scenarios described above, now 6 is working. But last 2(V2 relative paths) need bit more work. Only reason V3 works for relative path is if it can't determine file path type then it defaults to V3.
I'll check 1-2 more ideas if I can make V2 relative paths work, then I'll talk with you.

nkolev92 commented:
My best guess is that the source + weird error message about `supporting all` are orthogonal. 

One might prevent you from hitting the other, but if that's the case you should solve the problems in order. 
If you think there are multiple problems, solve them individually and PR them one by one.

erdembayar commented:
> My best guess is that the source + weird error message about `supporting all` are orthogonal.
> 
> One might prevent you from hitting the other, but if that's the case you should solve the problems in order.
> If you think there are multiple problems, solve them individually and PR them one by one.

Main problem is when we pass "..\Mysource" it doesn't know it was local source type then it default to [false](https://github.com/NuGet/NuGet.Client/blob/3ef6cd3c0145e59314a495569c2d53296794fbf3/src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSource.cs#L93). Maybe we can add step to check if Dir.Exists("..\Mysource") then make it `local = true`. That will solve relative path problem.

nkolev92 commented:
@erdembayar 

Look into https://github.com/NuGet/NuGet.Client/blob/3ef6cd3c0145e59314a495569c2d53296794fbf3/src/NuGet.Core/NuGet.Build.Tasks/BuildTasksUtility.cs#L703. 

We already do that in restore...if we're doing it different in other parts of the code, restore contains a tried and proven concept.

erdembayar committed:
Fix relative path is not discovered when resolving package from local.

erdembayar committed:
Add more unit test covering all possible scenarios.

erdembayar commented:
> @erdembayar
> 
> Look into
> 
> https://github.com/NuGet/NuGet.Client/blob/3ef6cd3c0145e59314a495569c2d53296794fbf3/src/NuGet.Core/NuGet.Build.Tasks/BuildTasksUtility.cs#L703
> 
> .
> We already do that in restore...if we're doing it different in other parts of the code, restore contains a tried and proven concept.

Ok. Fixed all possible paths and added 16 unit tests for covering them.

erdembayar committed:
Fix Linux unit tests.

erdembayar commented:
@nkolev92  @zkat 
Please review again.

erdembayar commented:
@NuGet/nuget-client please review.

nkolev92 commented:
Just wanted to make sure the comment doesn't get missed, see https://github.com/NuGet/NuGet.Client/pull/3783#discussion_r543841448.

erdembayar commented:
> Just wanted to make sure the comment doesn't get missed, see [#3783 (comment)](https://github.com/NuGet/NuGet.Client/pull/3783#discussion_r543841448).

I replied, can you check?

erdembayar committed:
Remove unneeded changes.

erdembayar committed:
Add passed sources into dgSpec sources.

erdembayar committed:
Refactor code: Instead of add source we replace source in dgSpec file.

erdembayar commented:
> Just wanted to make sure the comment doesn't get missed, see [#3783 (comment)](https://github.com/NuGet/NuGet.Client/pull/3783#discussion_r543841448).

@nkolev92 
Please check now. 
Instead of adding source now replacing them and wrote several unit test to check that behavior.

nkolev92 commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/AddPackageReferenceCommandRunner.cs:
```diff
@@ -61,6 +61,17 @@ public async Task<int> ExecuteCommand(PackageReferenceArgs packageReferenceArgs,
+            {
+                var projectDirectory = Path.GetDirectoryName(packageReferenceArgs.ProjectPath);
+
+                for (int i = 0; i < packageReferenceArgs.Sources.Length; i++)
```
for readability purposes consider `foreach`.

nkolev92 commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/AddPackageReferenceCommandRunner.cs:
```diff
@@ -61,6 +61,17 @@ public async Task<int> ExecuteCommand(PackageReferenceArgs packageReferenceArgs,
+
+                for (int i = 0; i < packageReferenceArgs.Sources.Length; i++)
+                {
+                    packageReferenceArgs.Sources[i] = UriUtility.GetAbsolutePath(projectDirectory, packageReferenceArgs.Sources[i]);
```
Personally I think we should avoid reassigning to the args.

The `args` would ideally be an immutable group of options/requests, ideally something that's reproducible.

erdembayar replied:
created new variable.

nkolev92 commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/AddPackageCommandUtility.cs:
```diff
@@ -116,5 +117,19 @@ public static List<PackageSource> EvaluateSources(IList<PackageSource> requested
+            if (newSources?.Any() == true)
+            {
+                packageSpec.RestoreMetadata.Sources =
+                    newSources.Where(ns => !string.IsNullOrEmpty(ns)).ToHashSet(StringComparer.OrdinalIgnoreCase).Select(ns => new PackageSource(ns)).ToList();
```
Do we need the hashset? 
Seems like an overkill given that the rest of the code likely handles it and it's an uncommon scenario. 

Or did you already hit a problem with dup sources and my assumption is inaccurate.

erdembayar replied:
Removed hashset.

nkolev92 commented on src/NuGet.Core/NuGet.Protocol/LocalRepositories/FindLocalPackagesResourceV3Provider.cs:
```diff
@@ -18,8 +18,10 @@ public FindLocalPackagesResourceV3Provider()
+            var feedType = await source.GetFeedType(token);
 
-            if (await source.GetFeedType(token) == FeedType.FileSystemV3)
+            if (feedType == FeedType.FileSystemV3
```
What does this do? 

Feels like you are trying to fix 2 problems in 1 PR here.

erdembayar replied:
Ok. Reverting this code.
Maybe better to create separate issue.

nkolev92 reviewed (commented):
I'd separate the `--source` fix from the other fixes.

erdembayar committed:
Revert other feed provider problem fix which need to go in seperate PR.

erdembayar committed:
Address code review comments by Nikolche.

erdembayar reviewed (commented):
@nkolev92 

Addressed PR review comments.

nkolev92 commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/AddPackageReferenceCommandRunner.cs:
```diff
@@ -102,6 +102,29 @@ public async Task<int> ExecuteCommand(PackageReferenceArgs packageReferenceArgs,
+
+                foreach (string source in packageReferenceArgs.Sources)
+                {
+                    sourcePaths.Add(UriUtility.GetAbsolutePath(projectDirectory, source));
```
I think this should be relative to the startup directory and not the project directory.

Try out how restore works, but I'm pretty confident that `dotnet restore --source` will work relative to the invocation directory.

erdembayar replied:
@nkolev92 
It looks current implementation is correct for now, because current stable dotnet (sdk 5.0.0 and 5.0.1) restore works with source relative to project, not current invocation directory. Even "5.0.100-preview.8.20417.9" work relative to project location.
Please check below output:

```
C:\Users\eryondon\Desktop\111\222> dotnet restore C:\NuGetProj\IssueRepro\5127\App2\App2.csproj /p:restoreSources="..\Mysource2"
  Determining projects to restore...
C:\Program Files\dotnet\sdk\5.0.101\NuGet.targets(131,5): error : The local source 'C:\NuGetProj\IssueRepro\5127\Mysource2' doesn't exist. [C:\NuGetProj\IssueRepro\5127\App2\App2.csproj]
```

But starting from `5.0.200-preview.20614.14` it checks from relative to current invocation directory. Not sure if it's permanent change or not. How do we handle this kind of breaking changes?
```
C:\Users\eryondon\Desktop\111\222> dotnet restore C:\NuGetProj\IssueRepro\5127\App2\App2.csproj /p:restoreSources="..\Mysource2"
  Determining projects to restore...
C:\Program Files\dotnet\sdk\5.0.200-preview.20614.14\NuGet.targets(131,5): error : The local source 'C:\Users\eryondon\Desktop\111\Mysource2' doesn't exist. [C:\NuGetProj\IssueRepro\5127\App2\App2.csproj]
```

nkolev92 replied:
That's because of https://github.com/NuGet/Home/issues/9406 which was fixed in https://github.com/NuGet/NuGet.Client/pull/3722. 

5.0.100 has a bug, 5.0.200 preview has the fix. 
5.0.200 is the one with the correct behavior.

erdembayar replied:
Ok.
I changed to `Environment.CurrentDirectory` to get StartUp directory. So far I tested with patched dotnet sdk from command line it works fine from any StartUp directory and source relative to it.

nkolev92 commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/XPlatAddPkgTests.cs:
```diff
@@ -745,6 +747,716 @@ public async Task AddPkg_ConditionalAddWithUserInputFramework_Success(string pac
+                var commandRunner = new AddPackageReferenceCommandRunner();
+
+                // Act
+                var result = commandRunner.ExecuteCommand(packageArgs, MsBuild)
```
you should probably await this instead of calling `.Result`

erdembayar replied:
Done.

erdembayar replied:
Ok. Resolved.

nkolev92 commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/XPlatAddPkgTests.cs:
```diff
@@ -745,6 +747,716 @@ public async Task AddPkg_ConditionalAddWithUserInputFramework_Success(string pac
+                Assert.Equal(0, result);
+
+                // Make sure source is replaced in generated dgSpec file.
+                PackageSpec[] packageSpecs = XPlatTestUtils.GetPackageSpecs(projectA);
```
you don't need to read the dg spec...just follow `projectA.AssetsFile.PackageSpec` instead.

erdembayar replied:
Ok. Addressed.

nkolev92 commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/XPlatAddPkgTests.cs:
```diff
@@ -745,6 +747,716 @@ public async Task AddPkg_ConditionalAddWithUserInputFramework_Success(string pac
         }
 
+        [Fact]
+        public async Task AddPkg_WithV2AdditionalSourceFeed_AbsolutePath_NoVersionSpecified_Success()
```
Either source or feed, not both :)

erdembayar replied:
Removed `feed` part.

erdembayar replied:
Removed `feed` part.

nkolev92 reviewed (commented):
Looks better. 
A couple of questions. 

The one thing I'm not confident about is that this fixes the actual issue it's referencing. 

I think that the `--source` support broke after the issue in question was filled. 

Did you ever get to the bottom of it? I wouldn't be shocked if it's not reproable anymore, but I want to make sure we're communicating the right thing to the customers.

erdembayar commented:
> The one thing I'm not confident about is that this fixes the actual issue it's referencing.
> 
@nkolev92 

I'm not sure what `what referencing` are you referring? Can you be more specific? Do you mean reference line in csproj file?

> I think that the `--source` support broke after the issue in question was filled.
> 

Can you be more specific? Exactly which `--source support` broken? Is there another one besides this `dotnet add package xxx -s 
 Sourcepath`? 
 
> Did you ever get to the bottom of it? I wouldn't be shocked if it's not reproable anymore, but I want to make sure we're communicating the right thing to the customers.

I'm not aware of any other issue besides this one. Can you give me link for other issue you're referrring?

erdembayar committed:
Address latest code review comment

nkolev92 commented:
> I'm not sure what what referencing are you referring? Can you be more specific? Do you mean reference line in csproj file?

You are referencing https://github.com/NuGet/Home/issues/5127, in your issue by saying: 

> Fixes: Issue#5127

You started off by working on https://github.com/NuGet/Home/issues/5127, but I'm not sure you ever got to the bottom of what all the customers were seeing. 

Another related issue was found (--source not working), and the issue was completely retitled to capture only that. 

Because it's been a long time since the original issue was filled, it doesn't seem like the customers are only hitting the `--source` issue, but they were hitting something else. 

In particular I'm not sure if you ever got the bottom of why `error: Package '<PackageName>' is incompatible with 'all' frameworks in project '/path/to/project.csproj'.` is happening. 

I think it might be ok, to say, the add package --source is something we got out of that issue, but with the latest tooling we can't repro the OP. 
Seems not ideal to retitle the issue given that the OP is communicating a different problem.

erdembayar committed:
Instead of relative to project directory, now resolve sources relative to startup directory.

erdembayar reviewed (commented):
> You started off by working on [NuGet/Home#5127](https://github.com/NuGet/Home/issues/5127), but I'm not sure you ever got to the bottom of what all the customers were seeing.
> 
> Another related issue was found (--source not working), and the issue was completely retitled to capture only that.
> 

From my testing `--source` was working as of version sdk 3.1.404 and broke as of sdk 5.1.100.

> Because it's been a long time since the original issue was filled, it doesn't seem like the customers are only hitting the `--source` issue, but they were hitting something else.
> 
> In particular I'm not sure if you ever got the bottom of why `error: Package '<PackageName>' is incompatible with 'all' frameworks in project '/path/to/project.csproj'.` is happening.
> 
> I think it might be ok, to say, the add package --source is something we got out of that issue, but with the latest tooling we can't repro the OP.
> Seems not ideal to retitle the issue given that the OP is communicating a different problem.

Sorry for confusion, I restored original title for [Issue#5127](https://github.com/NuGet/Home/issues/5127). But keeping the title for current PR("Fix dotnet add package doesn't work with relative path source feed or when version is not specified
"). Because actual problem described by user was subproblem of this, because user specified `-s` in repro instruction.
Also over time depending on dotnet version this issue repro of `Package 'NameOfPackage' is incompatible with 'all' frameworks` keep evolved and causing more confusion. Original [Issue#5127](https://github.com/NuGet/Home/issues/5127) filed back in 2017 with `sdk 1.0.1` and since then repro condition changed depending of sdk versions.

1. For dotnet 1.x.x-3.1.x. user always get `Package 'NameOfPackage' is incompatible with 'all' frameworks` error regardless of whether we supplied source or version. It simply means package is not found for any cases.

C:\NuGetProj\Temp\111\333>dotnet --version
3.1.404

C:\NuGetProj\Temp\111\333>dotnet add C:\NuGetProj\IssueRepro\5127\App2\App2.csproj package Domain2 -s ..\Packages\Mysource2
  Determining projects to restore...
  Writing C:\Users\eryondon\AppData\Local\Temp\tmp135A.tmp
info : Adding PackageReference for package 'Domain2' into project 'C:\NuGetProj\IssueRepro\5127\App2\App2.csproj'.
info : Restoring packages for C:\NuGetProj\IssueRepro\5127\App2\App2.csproj...
info :   GET https://api.nuget.org/v3-flatcontainer/domain2/index.json
info :   NotFound https://api.nuget.org/v3-flatcontainer/domain2/index.json 441ms
error: Unable to find package Domain2. No packages exist with this id in source(s): ..\Packages\Mysource2, nuget.org
error: Package **'Domain2' is incompatible with 'all' frameworks in project** 'C:\NuGetProj\IssueRepro\5127\App2\App2.csproj'.

C:\NuGetProj\Temp\111\333>dotnet add C:\NuGetProj\IssueRepro\5127\App2\App2.csproj package Domain2
  Determining projects to restore...
  Writing C:\Users\eryondon\AppData\Local\Temp\tmp5557.tmp
info : Adding PackageReference for package 'Domain2' into project 'C:\NuGetProj\IssueRepro\5127\App2\App2.csproj'.
info : Restoring packages for C:\NuGetProj\IssueRepro\5127\App2\App2.csproj...
info :   GET https://api.nuget.org/v3-flatcontainer/domain2/index.json
info :   NotFound https://api.nuget.org/v3-flatcontainer/domain2/index.json 470ms
error: Unable to find package Domain2. No packages exist with this id in source(s): nuget.org
error: **Package 'Domain2' is incompatible with 'all' frameworks in project** 'C:\NuGetProj\IssueRepro\5127\App2\App2.csproj'.

C:\NuGetProj\Temp\111\333>dotnet add C:\NuGetProj\IssueRepro\5127\App2\App2.csproj package Domain2 -v 1.0.0
  Determining projects to restore...
  Writing C:\Users\eryondon\AppData\Local\Temp\tmpA6C9.tmp
info : Adding PackageReference for package 'Domain2' into project 'C:\NuGetProj\IssueRepro\5127\App2\App2.csproj'.
info : Restoring packages for C:\NuGetProj\IssueRepro\5127\App2\App2.csproj...
info :   GET https://api.nuget.org/v3-flatcontainer/domain2/index.json
info :   NotFound https://api.nuget.org/v3-flatcontainer/domain2/index.json 243ms
error: Unable to find package Domain2. No packages exist with this id in source(s): nuget.org
error: **Package 'Domain2' is incompatible with 'all' frameworks in project** 'C:\NuGetProj\IssueRepro\5127\App2\App2.csproj'.

2. For starting from sdk 5.0.100 user only get `Package 'NameOfPackage' is incompatible with 'all' frameworks` if user requested `-v` version otherwise it simply says ` There are no versions available for the package` instead.

C:\NuGetProj\Temp\111\333>dotnet --version
5.0.100

C:\NuGetProj\Temp\111\333>dotnet add C:\NuGetProj\IssueRepro\5127\App2\App2.csproj package Domain2
  Determining projects to restore...
  Writing C:\Users\eryondon\AppData\Local\Temp\tmp3136.tmp
info : Adding PackageReference for package 'Domain2' into project 'C:\NuGetProj\IssueRepro\5127\App2\App2.csproj'.
info :   GET https://api.nuget.org/v3/registration5-gz-semver2/domain2/index.json
info :   NotFound https://api.nuget.org/v3/registration5-gz-semver2/domain2/index.json 244ms
info :   GET https://api.nuget.org/v3/registration5-gz-semver2/domain2/index.json
info :   NotFound https://api.nuget.org/v3/registration5-gz-semver2/domain2/index.json 147ms
error: **There are no versions available for the package** 'Domain2'.

C:\NuGetProj\Temp\111\333>dotnet add C:\NuGetProj\IssueRepro\5127\App2\App2.csproj package Domain2 -s ..\Packages\Mysource2
  Determining projects to restore...
  Writing C:\Users\eryondon\AppData\Local\Temp\tmpEBD2.tmp
info : Adding PackageReference for package 'Domain2' into project 'C:\NuGetProj\IssueRepro\5127\App2\App2.csproj'.
info :   GET https://api.nuget.org/v3/registration5-gz-semver2/domain2/index.json
info :   NotFound https://api.nuget.org/v3/registration5-gz-semver2/domain2/index.json 397ms
info :   GET https://api.nuget.org/v3/registration5-gz-semver2/domain2/index.json
info :   NotFound https://api.nuget.org/v3/registration5-gz-semver2/domain2/index.json 150ms
error: **There are no versions available for the package** 'Domain2'.

C:\NuGetProj\Temp\111\333>dotnet add C:\NuGetProj\IssueRepro\5127\App2\App2.csproj package Domain2 -v 1.0.0
  Determining projects to restore...
  Writing C:\Users\eryondon\AppData\Local\Temp\tmp673A.tmp
info : Adding PackageReference for package 'Domain2' into project 'C:\NuGetProj\IssueRepro\5127\App2\App2.csproj'.
info : Restoring packages for C:\NuGetProj\IssueRepro\5127\App2\App2.csproj...
info :   GET https://api.nuget.org/v3-flatcontainer/domain2/index.json
info :   NotFound https://api.nuget.org/v3-flatcontainer/domain2/index.json 106ms
error: NU1101: Unable to find package Domain2. No packages exist with this id in source(s): nuget.org
error: **Package 'Domain2' is incompatible with 'all' frameworks in project** 'C:\NuGetProj\IssueRepro\5127\App2\App2.csproj'.

C:\NuGetProj\Temp\111\333>dotnet add C:\NuGetProj\IssueRepro\5127\App2\App2.csproj package Domain2 -s ..\Packages\Mysource2 -v 1.0.0
  Determining projects to restore...
  Writing C:\Users\eryondon\AppData\Local\Temp\tmpA6A6.tmp
info : Adding PackageReference for package 'Domain2' into project 'C:\NuGetProj\IssueRepro\5127\App2\App2.csproj'.
info : Restoring packages for C:\NuGetProj\IssueRepro\5127\App2\App2.csproj...
info :   GET https://api.nuget.org/v3-flatcontainer/domain2/index.json
info :   NotFound https://api.nuget.org/v3-flatcontainer/domain2/index.json 203ms
error: NU1101: Unable to find package Domain2. No packages exist with this id in source(s): ..\Packages\Mysource2, nuget.org
error: **Package 'Domain2' is incompatible with 'all' frameworks in project** 'C:\NuGetProj\IssueRepro\5127\App2\App2.csproj'.

I addressed your concern about `StartUp directory`, please check. I used `Environment.CurrentDirectory` to capture StartUp directory. Another solution would be we pass it from dotnet to here as argument, but seems not necessary. Please let me know if pass by argument is better option due to some cases I'm not aware of it.

erdembayar commented:
@nkolev92 
Is it correct approach for start up directory? Using Enviroment.CurrentDirectory? 
Should be pass as optional/required parameter here?

nkolev92 commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/AddPackageReferenceCommandRunner.cs:
```diff
@@ -102,6 +102,29 @@ public async Task<int> ExecuteCommand(PackageReferenceArgs packageReferenceArgs,
+
+                foreach (string source in packageReferenceArgs.Sources)
+                {
+                    sourcePaths.Add(UriUtility.GetAbsolutePath(Environment.CurrentDirectory, source));
```
`dotnet add package` works in a peculiar way. We can verify that it works correctly by adding an extra integration test. 

Some background first. 
Read https://microsoft.sharepoint.com/teams/NuGet/_layouts/OneNote.aspx?id=%2Fteams%2FNuGet%2FTeam%2FNuGet&wd=target%28Engineering.one%7C219AABAA-18B3-46F4-83E8-D3188D83DE22%2FWhere%28%21%5C%29%20is%20NuGet%3F%7CC79766FE-842B-467F-BA94-438E51F8523C%2F%29
onenote:https://microsoft.sharepoint.com/teams/NuGet/Team/NuGet/Engineering.one#Where(!)%20is%20NuGet&section-id={219AABAA-18B3-46F4-83E8-D3188D83DE22}&page-id={C79766FE-842B-467F-BA94-438E51F8523C}&end. 

* NuGet ships NuGet.Commandline.Xplat in the .NET SDK, and commands such as `dotnet add package` are implemented by forwarding certain arguments. 
* Basically, there's a parsing on `dotnet.exe` process side, and then that's forwarded to `NUGet.Commandline.Xplat` that right now, repeats it's own parsing. 
* Environment.CurrentDirectory is the working directory that's passed at process start-up. Meaning the directory being read here is the one for `NuGet.Commandline.Xplat`. I am *not* sure if that is forwarded by the `dotnet.exe` implementation, but I'd expect that it is. 

You should make sure that it's resolve *against* the original startup directory and not relative to something like `nuget.commandline.xplat`'s start-up directory. 

That being said, `Dotnet.Integration.Tests` are the ones that run `dotnet.exe`, so I think adding a test in there tests the relative path is a good idea.

erdembayar replied:
Ok. I added source relative path unit tests into `Dotnet.Integration.Test`.
But I added source absolute path unit tests into Xplat.FuncTest, because previous one is slow than this one and I don't want to make it more slow by adding tests that can run faster elsewhere.

nkolev92 reviewed (commented):
Right track, a suggestion for more extensive automated testing.

erdembayar committed:
Start moving dotnet relative source tests to dotnet integration tests.

erdembayar committed:
Move another dotnet relative source test into dotnet.integration.test

erdembayar committed:
Fix empty source passed situation for dotnet add package.

erdembayar committed:
Test additional frameworks

erdembayar committed:
Finish moving dotnet relative source unit tests to dotnet.integration.test

erdembayar committed:
Fix new unit test names.

erdembayar commented:
@nkolev92 
It's ready to review. It's already passing all tests, I just committed small changes. I added source relative path unit tests into Dotnet.Integration.Test.
But I added source absolute path unit tests into Xplat.FuncTest, because previous one is slow than this one and I don't want to make it more slow by adding tests that can run faster elsewhere.

nkolev92 commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/AddPackageReferenceCommandRunner.cs:
```diff
@@ -102,6 +102,22 @@ public async Task<int> ExecuteCommand(PackageReferenceArgs packageReferenceArgs,
             var originalPackageSpec = matchingPackageSpecs.FirstOrDefault();
 
+            // Convert relative path to absolute path if there is any
+            List<string> sourcePaths = new List<string>();
```
naming: sources, not sourcepaths. These can be http sources, not necessarily paths.

erdembayar replied:
Ok. Renamed to `sources`.

nkolev92 commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/AddPackageReferenceCommandRunner.cs:
```diff
@@ -334,7 +350,7 @@ public static async Task<NuGetVersion> GetLatestVersionAsync(PackageSpec origina
                     GlobalPackagesFolder = packageReferenceArgs.PackageDirectory,
                     PreLoadedRequestProviders = providers,
-                    Sources = packageReferenceArgs.Sources?.ToList()
+                    Sources = sourcePaths
```
tbh, you can cut this. It's not used anyways. We should probably mark it as not used.

erdembayar replied:
Added comment here for why we're not passing value down here.

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetSourcesTests.cs:
```diff
@@ -453,6 +461,372 @@ public void TestCommandInvalidArguments(string command, int badCommandIndex)
         }
 
+        [Theory]
+        [InlineData("net5.0")]
```
nit: Doesn't matter for now, but you don't need a theory if you only have 1 test case. Use a fact instead.

erdembayar replied:
Initially I was thinking about testing many frameworks as much possible then considering heavy taxing on testing system I just changed to one framework. 
Ok. I changed to `Fact`.

nkolev92 replied:
As a rule, you should only pivot on things that can affect the result of your test. 
None of the changes you are making affect compat, so it's ok to not pivot on a framework. 

Note that the focus should be correctness and completeness. 
The testing time is a secondary consideration and imo irrelevant unless the above 2 satisified.

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetSourcesTests.cs:
```diff
@@ -453,6 +461,372 @@ public void TestCommandInvalidArguments(string command, int badCommandIndex)
 
+        [Theory]
+        [InlineData("net5.0")]
+        public async Task AddPkg_WithV2_RelativePath_NoVersionSpecified_Success(string targetFrameworks)
```
This test is not in the right file. The file suggests sources tests not add package command tests.

nkolev92 replied:
name: withV2 what? :) 

Is the feed type relevant to the whole scenario? 
I think local vs http feed is the scenario you are trying to solve.

erdembayar replied:
I moved in to new `DotnetAddPackageTests.cs`
Also I renamed each tests to reflect they're local sourcefeed.

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetSourcesTests.cs:
```diff
@@ -453,6 +461,372 @@ public void TestCommandInvalidArguments(string command, int badCommandIndex)
+                var packageX_V1 = new PackageIdentity(packageX, new NuGetVersion("1.0.0"));
+                var packageX_V2 = new PackageIdentity(packageX, new NuGetVersion("2.0.0"));
+                var packageXPath = Path.Combine(pathContext.WorkingDirectory, "Custompackages");
+                var sourceRelativePath = RuntimeEnvironmentHelper.IsWindows ? "..\\..\\Custompackages" : "../../Custompackages";
```
nit: You can use `Path.Combine("..","..", custompackages)`

That way you don't need the env check, the runtime does it for you.

erdembayar replied:
Ok. Addressed it. I'm not sure what I was thinking when I add this.

nkolev92 replied:
> I'm not sure what I was thinking when I add this.

I think that about most of the code I've not written today :D

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetSourcesTests.cs:
```diff
@@ -453,6 +461,372 @@ public void TestCommandInvalidArguments(string command, int badCommandIndex)
+                var packageX = "packageX";
+                var packageY = "packageY";
+                var packageY_V1 = new PackageIdentity(packageY, new NuGetVersion("1.0.0"));
+                var packageXPath = Path.Combine(pathContext.WorkingDirectory, "Custompackages");
```
naming: I think should name this, custom source. The name suggest a path to the package, but it's really a path to the source.

nkolev92 replied:
Note that `sourceRelativePath` already follows that logic, so this is about being consistent.

erdembayar replied:
Ok. I renamed to `customSourcePath`. Please check.

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetSourcesTests.cs:
```diff
@@ -453,6 +461,372 @@ public void TestCommandInvalidArguments(string command, int badCommandIndex)
+
+                // Generate Package, but packageX is not in nuget feed.
+                await SimpleTestPackageUtility.CreateFolderFeedV2Async(
+                    packageXPath, //not using solution source folder
```
packageXPath, but Y is in the source? I'm not following.

erdembayar replied:
This code is for unhappy path(fail case). By mistake user placed `packageY` package in `packageX`'s custom location. 
Then try to restore `packageX` then it should fail.

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetSourcesTests.cs:
```diff
@@ -453,6 +461,372 @@ public void TestCommandInvalidArguments(string command, int badCommandIndex)
+                PackageSpec packageSpec = projectA.AssetsFile.PackageSpec;
+                string[] sources = packageSpec.RestoreMetadata.Sources.Select(s => s.Name).ToArray();
+                Assert.Equal(sources.Count(), 1);
+                Assert.True(sources[0].Contains("Custompackages"));
```
The value here should be the full path, so you can check it against the full path I think.

erdembayar replied:
Sure, I changed to full address.

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetSourcesTests.cs:
```diff
@@ -453,6 +461,372 @@ public void TestCommandInvalidArguments(string command, int badCommandIndex)
+            }
+        }
+
+        [Theory]
```
If you are going to write a theory, you can probably combine the test cases. 

For example, the theory data can be the version.
```cs
[InlineDate("")]
[InlineDate("1.0.0")]
```

erdembayar replied:
I changed code to 'Fact'. Since dotnet integration test are slow I don't want overburden with many input cases.

nkolev92 replied:
> I changed code to 'Fact'. Since dotnet integration test are slow I don't want overburden with many input cases.


I was suggesting you merge test cases into 1. I think having a test for scenarios with and without version makes sense.

nkolev92 reviewed (commented):
Minor changes/comments. 

A few suggestions to clean-up the code and make it easier to follow.

erdembayar committed:
Address PR review by Nkolche.

erdembayar reviewed (commented):
@nkolev92 
Ok. I addressed PR review.
Moved dotnet integration tests into new `DotnetAddPackageTests.cs` file.
Also in order to make sure I'm not breaking existing default behavior before my code changes I added new 2 tests which doesn't use custom source, instead use default solution source. 
`AddPkg_V2LocalSourceFeed_WithDefaultSolutiuonSource_NoVersionSpecified_Success`
`AddPkg_V3LocalSourceFeed_WithDefaultSolutiuonSource_NoVersionSpecified_Success`

erdembayar committed:
Remove unused `sources` parameter passed down.

erdembayar commented:
@NuGet/nuget-client 
It's ready to review.

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetAddPackageTests.cs:
```diff
@@ -0,0 +1,478 @@
+        }
+
+        [Fact]
+        public async Task AddPkg_V2LocalSourceFeed_WithRelativePath_NoVersionSpecified_Success()
```
Do you need both a V2 and a V3 test case?

erdembayar replied:
I thought V2 and V3 sources have different directory structure. So I thought testing both of them relative path would be cover all possible scenarios.
I can remove one of them. So we keep V3 one.

erdembayar replied:
@nkolev92 
I removed V2 source tests. Please check now.

nkolev92 replied:
> I thought V2 and V3 sources have different directory structure. So I thought testing both of them relative path would be cover all possible scenarios.

The reason why I kept probing around this topic is the fact that I wanted to make you are focusing on the right scenarios for the test. 

You needed to test 2 things: 

1. `--source` completely overwrites the source.
1. `--source` when used with a relative local source, the full path is calculated correctly.

The full path calculation is not affected by the feed type (v2 or v3 local feed), that only matters afterwards. 

I think the tests are there right now.
Taking another look :)

erdembayar committed:
Remove V2 sourcefeed unit tests

nkolev92 commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/XPlatTestUtils.cs:
```diff
@@ -219,14 +219,15 @@ internal static PackageReferenceArgs GetPackageReferenceArgs(string packageId, S
 
         internal static PackageReferenceArgs GetPackageReferenceArgs(string packageId, string packageVersion, SimpleTestProjectContext project,
-            string frameworks = "", string packageDirectory = "", string sources = "", bool noRestore = false, bool noVersion = false, bool prerelease = false)
+            string frameworks = "", string packageDirectory = "", string sources = "", bool noRestore = false, bool noVersion = false, bool prerelease = false, string dgFilePath = "")
```
Do you still need this file change?

erdembayar replied:
No need, I had some learning curve around Xplat env. I'll do better next time here. Reverted it.

nkolev92 reviewed (approved):
nit: watch the Pr title :) source and feed are the same thing.

erdembayar committed:
Revert unused file change.

erdembayar commented:
> This eyeballs fine to me. I would like to see tests specific to this scenario before approving, though.

@zkat 
Can you review now? I added unit tests.

erdembayar committed:
Merge branch 'dev' into dev-eryondon-package-incompatible-allframworks
